pkg/graveler/ref/manager.go

```diff
@@ -213,23 +271,95 @@ func (m *KVManager) FindMergeBase(ctx context.Context, repositoryID graveler.Rep
 	if err != nil {
 		return nil, err
 	}
-	return NewOrderedCommitIterator(ctx, m.db.db, repositoryID, IteratorPrefetchSize), nil
+	return NewKVOrderedCommitIterator(ctx, &m.kvStore, repositoryID, false)
 }
 
 func (m *KVManager) FillGenerations(ctx context.Context, repositoryID graveler.RepositoryID) error {
-	return m.db.FillGenerations(ctx, repositoryID)
+	// update commitNodes' generation in nodes "tree" using BFS algorithm.
+	// using a queue implementation
+	// adding a node to the queue only after all of its parents were visited in order to avoid redundant visits of nodesCommitIDs
+	nodes, err := m.createCommitIDsMap(ctx, repositoryID)
+	if err != nil {
+		return err
+	}
+	rootsCommitIDs := m.getRootNodes(nodes)
+	m.mapCommitNodesToChildren(nodes)
+	m.addGenerationToNodes(nodes, rootsCommitIDs)
+	return m.updateCommitGeneration(ctx, repositoryID, nodes)
+}
+
+func (m *KVManager) createCommitIDsMap(ctx context.Context, repositoryID graveler.RepositoryID) (map[graveler.CommitID]*CommitNode, error) {
+	iter, err := m.ListCommits(ctx, repositoryID)
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Close()
+
+	nodes := make(map[graveler.CommitID]*CommitNode)
+	for iter.Next() {
+		commit := iter.Value()
+		parentsToVisit := map[graveler.CommitID]struct{}{}
+		for _, parentID := range commit.Parents {
+			parentsToVisit[parentID] = struct{}{}
+		}
+
+		nodes[commit.CommitID] = &CommitNode{parentsToVisit: parentsToVisit}
+	}
+	if err := iter.Err(); err != nil {
+		return nil, err
+	}
+	return nodes, nil
+}
+
+func (m *KVManager) getRootNodes(nodes map[graveler.CommitID]*CommitNode) []graveler.CommitID {
+	var rootsCommitIDs []graveler.CommitID
+	for commitID, node := range nodes {
+		if len(node.parentsToVisit) == 0 {
+			rootsCommitIDs = append(rootsCommitIDs, commitID)
+		}
+	}
+	return rootsCommitIDs
+}
+
+func (m *KVManager) mapCommitNodesToChildren(nodes map[graveler.CommitID]*CommitNode) {
+	for commitID, commitNode := range nodes {
+		// adding current node as a child to all parents in commitNode.parentsToVisit
+		for parentID := range commitNode.parentsToVisit {
+			nodes[parentID].children = append(nodes[parentID].children, commitID)
+		}
+	}
+}
+
+func (m *KVManager) addGenerationToNodes(nodes map[graveler.CommitID]*CommitNode, rootsCommitIDs []graveler.CommitID) {
+	nodesCommitIDs := rootsCommitIDs
+	for currentGeneration := 1; len(nodesCommitIDs) > 0; currentGeneration++ {
+		var nextIterationNodes []graveler.CommitID
+		for _, nodeCommitID := range nodesCommitIDs {
+			currentNode := nodes[nodeCommitID]
+			nodes[nodeCommitID].generation = currentGeneration
+			for _, childNodeID := range currentNode.children {
+				delete(nodes[childNodeID].parentsToVisit, nodeCommitID)
+				if len(nodes[childNodeID].parentsToVisit) == 0 {
+					nextIterationNodes = append(nextIterationNodes, childNodeID)
+				}
+			}
+		}
+		nodesCommitIDs = nextIterationNodes
+	}
+}
+
+func (m *KVManager) Store() *kv.StoreMessage {
```

Comment:
Don't think it's really needed

Reply:
it is needed for initializing the commit iterator on the garbage collection

Reply:
See previous comment of @itaiad200 regarding the GC. Maybe we can just find a way to return the correct iterator from the ref manager service itself?

Reply:
done

Reply:
Not yet :)